### ISSUE TITLE
Transactions are difficult on shared connections

### ISSUE THREAD

**Issue by nugend** [omnilib/aiosqlite#19]
### Description

Currently, the library presumes that any sequential set of operations should be allowed to happen concurrently with any other set of operations. This is incompatible with transactional operations.

Here's a simple program to demonstrate the existing behavior and the desired behavior:

```
import aiosqlite                                                                                                 
import asyncio                                                                                                   
import os                                                                                                        
                                                                                                                 
async def insert_100_error(conn,val):                                                                            
    async with conn.cursor() as c:                                                                               
        await c.execute('BEGIN')                                                                                 
        await c.executemany("INSERT INTO foo VALUES (?)", [(val,) for x in range(100)])                          
        await c.execute("INSERT INTO foo VALUES (?)", (val + 'done',))                                           
        await c.execute('COMMIT')                                                                                
                                                                                                                 
async def insert_100_helper(conn,val):                                                                           
    def my_operation(c):                                                                                         
        c.execute('BEGIN')                                                                                       
        c.executemany("INSERT INTO foo VALUES (?)", [(val,) for x in range(100)])                                
        c.execute("INSERT INTO foo VALUES (?)", (val + 'done',))                                                 
        c.execute('COMMIT')                                                                                      
    async with conn.cursor() as c:                                                                               
        await c._execute(my_operation,c._cursor)                                                                 
                                                                                                                 
async def run():                                                                                                 
    async with aiosqlite.connect('test_linear.db') as conn:                                                      
        await conn.execute("CREATE TABLE foo (val1 text)")                                                       
        await conn.commit()                                                                                      
        await asyncio.gather(*[insert_100_error(conn,x) for x in ['a','b','c','d']])                             
        #await asyncio.gather(*[insert_100_helper(conn,x) for x in ['a','b','c','d']])                           
                                                                                                                 
                                                                                                                 
os.remove('test_linear.db')                                                                                      
asyncio.get_event_loop().run_until_complete(run())                                                                                                           
```

So rather than submit each individual function on a cursor to the sqlite thread, we submit a callback that takes a plain cursor which can be used directly.

I don't know if there are significant drawbacks to this approach, but it works the way I expect it to whereas I can't see a way to accomplish this with the library right now.

Not sure what you'd call such a function.

`Cursor.executelinear`? `executecb`? `executesync`?

If I've just missed the right way to do this with the current API, please let me know.

### Details

* OS: Linux
* Python version: 3.6
* aiosqlite version: v0.8
* Can you repro on master? Yes
* Can you repro in a clean virtualenv? Yes

**Comment by jreese** [comment]
This is mostly due to the way the underlying `sqlite3` module operates, such that it was designed for use by a signal thread – using `sqlite3` with libraries like gevent or greenlets would also result in similar transactional collisions.  Sending functions to execute with bundled queries is interesting, but otherwise limits the benefits of using asyncio, as you will need to wait for all queries in that bundle to complete before other queries can execute, even if they wouldn't otherwise be blocked on a transaction. 

My general suggestion would be that if you need to execute multiple queries simultaneously while preserving transaction safety, then you should be using multiple connection objects.  This is a common pattern with many asyncio database libraries that wrap synchronous connection libraries, like mysql, so this isn't necessarily specific to sqlite3 or aiosqlite.

**Comment by nugend** [comment]
This objection doesn't really make a lot of sense to me. The sqlite thread is going to pull queries off and execute them in the order they appear on the queue regardless of what's happening on the event loop thread. If I have sent two select queries from two concurrently executing tasks, one of those selects is going to execute first and the other will execute second.

As it is, adding such an additional function wouldn't change anything about the existing parallelism of operations using a single connection, it would just make it easier to write a transaction that behaved correctly.

**Comment by nugend** [comment]
Did you have any thoughts on my response? I'm happy to just monkeypatch the Cursror class in my own code to do as I wanted (it's literally one line). I just thought other folks who happen to use this might find it worthwhile.

**Comment by jreese** [comment]
Sorry for the delay in response. 

While I'm still skeptical of the necessity, given that I feel like anything depending on transactions should be using separate connections for each concurrent transaction (I'd feel the same for use of the standard sqlite module as well), I don't think the use case is unreasonable. I do believe it would at least be a performance improvement to run sequential queries on the dedicated thread, minimizing the amount of back and forth needed with the locks and queues, and we've added similar helpers for performance already. 

I'd be alright if we added something like a `.executefunction` method to both the connection and cursor classes, as long as it's well documented as an additional feature not available in sqlite3. In both cases, it seems reasonable to expect functions to take a cursor as the only parameter – the connection variant could create a new cursor to pass, while the cursor variant could just pass itself.

Feel free to submit a PR, and I'll do my best to review it in a more timely fashion. 😅
